protocol/contracts/TrucoMatch.sol

```diff
@@ -10,42 +10,132 @@ import "./crypt/DeckCrypt.sol";
 
 contract TrucoMatch  {
 
+    struct player {
+        address playerAddress;
+        uint256 tokensAtStake;
+    }
+    
     struct Match {
-        address[2] players; // player 0 is the creator of the match
+        player[2] players; // player 0 is the creator of the match
         CardsStructs.Deck deck;
         CardsStructs.GameState gameState;
         uint256 tokensAtStake; 
+
     }
 
     IERC3333 trucoEngine;
     DeckCrypt deckCrypt;
-    IERC20 trucoin;
+    IERC20 truCoin;
     Match currentMatch;
 
-    constructor (IERC3333 _trucoEngine, DeckCrypt _deckCrypt, IERC20 _trucoin) {
+    // Events
+    event MatchCreated(address match_address);
+    event MatchStarted(address player1, address player2, uint256 tokensAtStake);
+    event PlayerStaked(address player, uint256 tokensAtStake);
+    event DealStarted(address shuffler);
+    event DealEnded();
+
+
+    constructor (IERC3333 _trucoEngine, DeckCrypt _deckCrypt, IERC20 _truCoin, uint256 _tokensAtStake) {
+
         trucoEngine = _trucoEngine;
         deckCrypt = _deckCrypt;
-        currentMatch.players[0] = msg.sender;
+        truCoin = _truCoin;
+        currentMatch.tokensAtStake = _tokensAtStake;
+        currentMatch.players[0].playerAddress = msg.sender;
+
+        //Mint Sould Bound Token
 
-        //Transfer Trucoins from owner to contract
-        //Mint Sould Bound Token 
+        emit MatchCreated(address(this));
     }
-    
         
-    // Method for joining the match 
+    // Method for second player joining the match 
     function join() public {
```

Comment:
Calling `join()` implies calling `stake`, so are we checking preconditions twice?

Reply:
I divided join and stake cause first user is joined at constructor, so he don't need to join, but need to stake/bet. But I should deleted the requires from join method that are already at stake function